src/equations/shallow_water_wet_dry_2d.jl

```diff
@@ -0,0 +1,844 @@
+    \frac{\partial}{\partial t}(h v_1) + \frac{\partial}{\partial x}\left(h v_1^2 + \frac{g}{2}h^2\right)
+    + \frac{\partial}{\partial y}(h v_1 v_2) + g h \frac{\partial b}{\partial x} &= 0 \\
+    \frac{\partial}{\partial t}(h v_2) + \frac{\partial}{\partial x}(h v_1 v_2)
+    + \frac{\partial}{\partial y}\left(h v_2^2 + \frac{g}{2}h^2\right) + g h \frac{\partial b}{\partial y} &= 0.
+\end{aligned}
+```
+The unknown quantities of the SWE are the water height ``h`` and the velocities ``\mathbf{v} = (v_1, v_2)^T``.
+The gravitational constant is denoted by `g` and the (possibly) variable bottom topography function ``b(x,y)``.
+Conservative variable water height ``h`` is measured from the bottom topography ``b``, therefore one
+also defines the total water height as ``H = h + b``.
+
+The additional quantity ``H_0`` is also available to store a reference value for the total water height that
+is useful to set initial conditions or test the "lake-at-rest" well-balancedness.
+
+Also, there are two thresholds which prevent numerical problems as well as instabilities. Both of them do not
+have to be passed, as default values are defined within the struct. The first one, `threshold_limiter`, is
+used in [`PositivityPreservingLimiterShallowWater`](@ref) on the water height, as a (small) shift on the initial
+condition and cutoff before the next time step. The second one, `threshold_wet`, is applied on the water height to
+define when the flow is "wet" before calculating the numerical flux.
+
+The bottom topography function ``b(x,y)`` is set inside the initial condition routine
+for a particular problem setup. To test the conservative form of the SWE one can set the bottom topography
+variable `b` to zero.
+
+In addition to the unknowns, TrixiShallowWater.jl currently stores the bottom topography values at the approximation points
+despite being fixed in time. This is done for convenience of computing the bottom topography gradients
+on the fly during the approximation as well as computing auxiliary quantities like the total water height ``H``
+or the entropy variables.
+This affects the implementation and use of these equations in various ways:
+* The flux values corresponding to the bottom topography must be zero.
+* The bottom topography values must be included when defining initial conditions, boundary conditions or
+  source terms.
+* [`AnalysisCallback`](@ref) analyzes this variable.
+* Trixi.jl's visualization tools will visualize the bottom topography by default.
+
+References for the SWE are many but a good introduction is available in Chapter 13 of the book:
+- Randall J. LeVeque (2002)
+  Finite Volume Methods for Hyperbolic Problems
+  [DOI: 10.1017/CBO9780511791253](https://doi.org/10.1017/CBO9780511791253)
+"""
+struct ShallowWaterEquationsWetDry2D{RealT <: Real} <:
+       Trixi.AbstractShallowWaterEquations{2, 4}
+    gravity::RealT # gravitational constant
+    H0::RealT      # constant "lake-at-rest" total water height
```

Comment:
I don't really understand this - why do you still store the gravity_constant and H0? Aren't they already part of `basic_swe`?

At least to me it seems counter-intuitive to have them stored in two different places if they are the same, and underdocumented if they are actually different

Reply:
Even though we could access these variables from the `basic_swe`, I would like to keep the duplicates for convenience (a similar approach was used in Trixi.jl for the [Navier-Stokes equations](https://github.com/trixi-framework/Trixi.jl/blob/main/src/equations/compressible_navier_stokes_1d.jl#L104). 
Accessing these variables from the `basic_swe` would require use to call via `equations.basic_swe.gravity_constant` and therefore reduce readability.
Furthermore, the variables are then passed to a struct outside of `TrixiShallowWater.jl`, such that they are not directly visible for developers in `TrixiShallowWater.jl`

To explain the motivation behind this, I have added a comment in 42cf9eeb711b87eaee5b28401b115e2ae99229ff.

Reply:
Sounds good to me. Thanks for the excellent preparation of the reasoning, incl the reference to the NSE 💪